llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,272 @@
+//===========- DynamicTableEmitter.cpp - Generate dynamic tables -============//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This tablegen backend emits a key-value map that can be dynamically extended
+// at runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/TableGen/Error.h"
+#include "llvm/TableGen/Record.h"
+#include "llvm/TableGen/TableGenBackend.h"
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "dynamic-table-emitter"
+
+namespace {
+
+int getAsInt(Init *B) {
+  return cast<IntInit>(
+             B->convertInitializerTo(IntRecTy::get(B->getRecordKeeper())))
+      ->getValue();
+}
+
+struct GenericField {
+  std::string Name;
+  RecTy *RecType = nullptr;
+  bool IsCode = false;
+  bool IsList = false;
+
+  GenericField(StringRef Name) : Name(std::string(Name)) {}
+};
+
+struct DynamicTable {
+  std::string Name;
+  ArrayRef<SMLoc> Locs; // Source locations from the Record instance.
+  std::string PreprocessorGuard;
+  std::string CppTypeName;
+  SmallVector<GenericField, 2> Fields;
+  std::vector<Record *> Entries;
+
+  const GenericField *getFieldByName(StringRef Name) const {
+    for (const auto &Field : Fields) {
+      if (Name == Field.Name)
+        return &Field;
+    }
+    return nullptr;
+  }
+};
+
+class DynamicTableEmitter {
+  RecordKeeper &Records;
+  std::set<std::string> PreprocessorGuards;
+
+public:
+  DynamicTableEmitter(RecordKeeper &R) : Records(R) {}
+
+  void run(raw_ostream &OS);
+
+private:
+  std::string primaryRepresentation(SMLoc Loc, const GenericField &Field,
+                                    Init *I) {
+    if (StringInit *SI = dyn_cast<StringInit>(I)) {
+      if (Field.IsCode || SI->hasCodeFormat())
+        return std::string(SI->getValue());
+
+      return SI->getAsString();
+    } else if (BitsInit *BI = dyn_cast<BitsInit>(I))
+      return "0x" + utohexstr(getAsInt(BI));
+    else if (BitInit *BI = dyn_cast<BitInit>(I))
+      return BI->getValue() ? "true" : "false";
+    else if (Field.IsList) {
+      if (auto LI = dyn_cast<ListInit>(I)) {
+        std::stringstream res;
+        auto values = LI->getValues();
+        bool IsAspect = (Field.Name == "aspects");
+        for (const auto &[idx, val] : enumerate(values)) {
+          if (idx > 0)
+            res << ", ";
+
```

Comment:
I've just discovered `ListSeparator` which you used in another method and I think that we should use it here, it would simplify the code